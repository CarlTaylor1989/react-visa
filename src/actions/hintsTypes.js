export const SET_CURRENT_HINT_GROUP = 'SET_CURRENT_HINT_GROUP';
export const SET_HINT_GROUP_INDEX = 'SET_HINT_GROUP_INDEX';
export const SET_HINTS_PAUSED = 'SET_HINTS_PAUSED';
