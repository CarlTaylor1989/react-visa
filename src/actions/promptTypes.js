export const ADD_PRODUCT_PROMPT = 'ADD_PRODUCT_PROMPT';
export const RESET_PRODUCT_PROMPT = 'RESET_PRODUCT_PROMPT';
export const ADD_BONUS_STREAK_PROMPT = 'ADD_BONUS_STREAK_PROMPT';
export const RESET_BONUS_STREAK_PROMPT = 'RESET_BONUS_STREAK_PROMPT';
export const ADD_ACHIEVEMENT_PROMPT = 'ADD_ACHIEVEMENT_PROMPT';
export const RESET_ACHIEVEMENT_PROMPTS = 'RESET_ACHIEVEMENT_PROMPTS';
export const ADD_RANK_PROMPT = 'ADD_RANK_PROMPT';
export const RESET_RANK_PROMPT = 'RESET_RANK_PROMPT';
export const ADD_PROMPTS = 'ADD_PROMPTS';
export const SET_PROMPTS_PAUSED = 'SET_PROMPTS_PAUSED';
export const SET_DISPLAY_PROMPTS = 'SET_DISPLAY_PROMPTS';
export const RESET_DISPLAYED_PROMPTS = 'RESET_DISPLAYED_PROMPTS';
export const RESET_PROMPTS = 'RESET_PROMPTS';
