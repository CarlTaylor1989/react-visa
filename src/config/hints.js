export const HINT_DIAGNOSTIC = 'diagnostic';
export const HINT_NETWORK = 'network';
export const HINT_PRODUCT = 'product';
export const HINT_CLIENT_SELECTION = 'clientselection';
export const HINT_CLIENT_PROPOSAL = 'clientproposal';

export const HINT_TIME = 20000;

export default [
  {
    id: 'h1',
    group: HINT_DIAGNOSTIC
  },
  {
    id: 'h2',
    group: HINT_NETWORK
  },
  {
    id: 'h3',
    group: HINT_NETWORK
  },
  {
    id: 'h4',
    group: HINT_NETWORK
  },
  {
    id: 'h5',
    group: HINT_NETWORK
  },
  {
    id: 'h6',
    group: HINT_NETWORK
  },
  {
    id: 'h7',
    group: HINT_NETWORK
  },
  {
    id: 'h8',
    group: HINT_NETWORK
  },
  {
    id: 'h9',
    group: HINT_PRODUCT
  },
  {
    id: 'h10',
    group: HINT_PRODUCT
  },
  {
    id: 'h11',
    group: HINT_CLIENT_SELECTION
  },
  {
    id: 'h12',
    group: HINT_CLIENT_PROPOSAL
  },
  {
    id: 'h13',
    group: HINT_CLIENT_PROPOSAL
  }
];
