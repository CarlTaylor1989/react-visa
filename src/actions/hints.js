import * as types from './hintsTypes';

export function setCurrentHintGroup(group) {
  return {
    type: types.SET_CURRENT_HINT_GROUP,
    group
  };
}

export function setHintGroupIndex(hints) {
  return {
    type: types.SET_HINT_GROUP_INDEX,
    hints
  };
}

export function setHintsPaused(paused) {
  return {
    type: types.SET_HINTS_PAUSED,
    paused
  };
}
