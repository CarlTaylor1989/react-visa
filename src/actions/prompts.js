import * as types from './promptTypes';

export function addProductPrompt(id) {
  return {
    type: types.ADD_PRODUCT_PROMPT,
    id
  };
}

export function resetProductPrompt() {
  return {
    type: types.RESET_PRODUCT_PROMPT
  };
}

export function addBonusStreakPrompt() {
  return {
    type: types.ADD_BONUS_STREAK_PROMPT
  };
}

export function resetBonusStreakPrompt() {
  return {
    type: types.RESET_BONUS_STREAK_PROMPT
  };
}

export function addAchievementPrompt(id) {
  return {
    type: types.ADD_ACHIEVEMENT_PROMPT,
    id
  };
}

export function resetAchievementPrompts() {
  return {
    type: types.RESET_ACHIEVEMENT_PROMPTS
  };
}

export function addRankPrompt(id) {
  return {
    type: types.ADD_RANK_PROMPT,
    id
  };
}

export function resetRankPrompt() {
  return {
    type: types.RESET_RANK_PROMPT
  };
}

export function addPrompts(prompts) {
  return {
    type: types.ADD_PROMPTS,
    prompts
  };
}

export function setPromptsPaused(paused) {
  return {
    type: types.SET_PROMPTS_PAUSED,
    paused
  };
}

export function setDisplayPrompts() {
  return {
    type: types.SET_DISPLAY_PROMPTS
  };
}

export function resetDisplayedPrompts() {
  return {
    type: types.RESET_DISPLAYED_PROMPTS
  };
}

export function resetPrompts() {
  return {
    type: types.RESET_PROMPTS
  };
}
