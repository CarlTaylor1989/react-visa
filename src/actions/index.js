import { push } from 'connected-react-router';
import * as AchievementActions from './achievements';
import * as AudioActions from './audio';
import * as ClientActions from './client';
import * as DiagnosticActions from './diagnostic';
import * as GenericActions from './generic';
import * as HintsActions from './hints';
import * as MapActions from './map';
import * as PromptActions from './prompts';
import * as ScoreActions from './score';
import * as ScormActions from './scorm';
import * as SettingsActions from './settings';
import * as StreakActions from './streak';
import * as TutorialActions from './tutorial';

const ActionCreators = Object.assign(
  {},
  AchievementActions,
  AudioActions,
  ClientActions,
  DiagnosticActions,
  GenericActions,
  HintsActions,
  MapActions,
  PromptActions,
  ScoreActions,
  ScormActions,
  SettingsActions,
  StreakActions,
  TutorialActions,
  { push }
);

export default ActionCreators;
