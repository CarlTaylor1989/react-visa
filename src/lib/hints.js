import hints, * as lib from '../config/hints';

/**
 * Filters groups by group into a new array.
 * @param {array} group Array of hints
 * @return {array}
 */
export const getHints = group => hints.filter(hint => hint.group === group);

/**
 * Gets the hint id for a hint based on it's group and index within that group.
 * @param {string} group Group name
 * @param {number} index Current index
 * @return {string | null}
 */
export const getHintId = (group, index) => {
  const hint = getHints(group)[index];
  return hint ? hint.id : null;
};

/**
 * Calculates whether a network hint is accessible.
 * @param {string} hintId Hint identifier
 * @param {object} state Game state
 * @return {boolean}
 */
export const networkHintAccessible = (hintId, state) => {
  let accessible = false;
  switch (hintId) {
    case 'h2':
      accessible = state.completedChallenges.length === 0;
      break;
    case 'h3':
    case 'h4':
    case 'h5':
    case 'h6':
      accessible = state.completedChallenges.length > 0;
      break;
    case 'h7':
      accessible = state.completedProducts.length > 0;
      break;
    case 'h8':
      accessible = state.bonusStreak === 3;
      break;
    default:
      accessible = true;
      break;
  }

  return accessible;
};

/**
 * Calculates whether a product hint is accessible.
 * @param {string} hintId Hint identifier
 * @param {object} state Game state
 * @return {boolean}
 */
export const productHintAccessible = (hintId, state) => {
  let accessible = false;
  switch (hintId) {
    case 'h9':
      accessible = state.completedChallenges.length > 0;
      break;
    case 'h10':
      accessible = state.productPageInactive;
      break;
    default:
      accessible = true;
      break;
  }
  return accessible;
};

export const clientProposalHintAccesible = (hintId, state) => {
  let accessible = false;
  const {
    clients,
    clientShowing,
    displayReqFeedback
  } = state;
  const {
    reqCorrect
  } = clients.find(({ id }) => id === clientShowing);
  switch (hintId) {
    case 'h12':
      accessible = !reqCorrect || (reqCorrect && displayReqFeedback);
      break;
    case 'h13':
      accessible = (reqCorrect && !displayReqFeedback);
      break;
    default:
      accessible = true;
      break;
  }
  return accessible;
};

/**
 * Calculates whether a hint is accessible based on game state. Custom logic for many of the hints.
 * @param {string} group Group name
 * @param {string} hintId Hint identifier
 * @param {object} state Game state
 * @return {boolean}
 */
export const hintAccessible = (group, hintId, state) => {
  let accessible = false;
  if (typeof hintId === 'string') {
    switch (group) {
      case lib.HINT_NETWORK:
        accessible = networkHintAccessible(hintId, state);
        break;
      case lib.HINT_PRODUCT:
        accessible = productHintAccessible(hintId, state);
        break;
      case lib.HINT_CLIENT_PROPOSAL:
        accessible = clientProposalHintAccesible(hintId, state);
        break;
      case lib.HINT_DIAGNOSTIC:
      case lib.HINT_CLIENT_SELECTION:
      default:
        accessible = true;
        break;
    }
  }
  return accessible;
};

/**
 * Recursive function that finds the next valid array index from the
 * array of hints filtered by group, this function takes note of the starting index
 * and returns it if it is reached again as to not end up in a infinite loop of recursive
 * function calls if no hints are currently available for a hint group.
 *
 * @param {string} group Group name
 * @param {number} index Current index
 * @param {object} state Game state
 * @param {number} startingIndex Starting index
 * @return {number}
 */
const getNextHintIndexRecursive = (group, index, state, startingIndex) => {
  const size = getHints(group).length - 1;
  const nextIndex = (index + 1 > size) ? 0 : index + 1;
  const hintId = getHintId(group, nextIndex);
  return (hintAccessible(group, hintId, state) || nextIndex === startingIndex)
    ? nextIndex
    : getNextHintIndexRecursive(group, nextIndex, state, startingIndex);
};

/**
 * Function to find the next valid array index from the
 * array of hints filtered by group.
 * @param {string} group Group name
 * @param {number} index Current index
 * @param {object} state Game state
 * @return {number}
 */
export const getNextHintIndex = (group, index, state) => getNextHintIndexRecursive(
  group,
  index,
  state,
  index
);

/**
 * Gets the current hint index for the current hint group from
 * game state.
 * @param {array} hintsState Array of hints from game state
 * @param {string} group Group name from game state
 * @return {number}
 */
export const getHintIndexFromState = (hintsState, group) => {
  const currentGroup = hintsState.find(hint => hint.group === group);
  return (currentGroup) ? currentGroup.index : null;
};
