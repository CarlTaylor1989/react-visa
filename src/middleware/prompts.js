import _ from 'lodash';
import * as promptTypes from '../actions/promptTypes';

export const addPrompts = () => next => (action) => {
  const updatedAction = action;
  if (updatedAction.type === promptTypes.ADD_PROMPTS) {
    if (!_.isArray(updatedAction.prompts.achievements)) {
      updatedAction.prompts.achievements = [];
    }
    if (!_.isArray(updatedAction.prompts.products)) {
      updatedAction.prompts.products = [];
    }
    if (!_.isArray(updatedAction.prompts.ranks)) {
      updatedAction.prompts.ranks = [];
    }
  }
  next(updatedAction);
};

export default [
  addPrompts
];
