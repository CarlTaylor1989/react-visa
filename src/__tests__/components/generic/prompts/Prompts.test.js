/* global toJson */
import React from 'react'; // eslint-disable-line no-unused-vars
import {
  Prompts,
  mapStateToProps,
  mapDispatchToProps
} from '../../../../components/generic/prompts/Prompts';
import * as hintsActions from '../../../../actions/hints';
import * as promptsActions from '../../../../actions/prompts';
import * as hintTypes from '../../../../actions/hintsTypes';
import * as promptsTypes from '../../../../actions/promptTypes';
import {
  ACHIEVEMENT_PROMPT,
  PRODUCT_PROMPT,
  RANK_PROMPT,
  BONUS_STREAK_PROMPT
} from '../../../../config/constants';

let mockProps;
beforeEach(() => {
  mockProps = {
    achievements: [],
    bonusStreak: 0,
    categories: {
      achievements: 0,
      products: 1
    },
    displayPrompts: false,
    paused: false,
    productFamily: 'm1f2',
    productMap: 'm1',
    products: [],
    ranks: [],
    resetAchievementPrompts: jest.fn(),
    resetBonusStreakPrompt: jest.fn(),
    resetDisplayedPrompts: jest.fn(),
    resetProductPrompt: jest.fn(),
    resetRankPrompt: jest.fn(),
    setDisplayPrompts: jest.fn(),
    setHintsPaused: jest.fn(),
    setPlayAudio: jest.fn()
  };
});

test('renders a closed popup', () => {
  const wrapper = shallow(<Prompts {...mockProps} />, {
    disableLifecycleMethods: true,
  });
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

describe('method componentDidMount()', () => {
  test('prompts are not set', () => {
    const wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().setPrompts = jest.fn();
    wrapper.instance().componentDidMount();
    expect(wrapper.instance().setPrompts).not.toHaveBeenCalled();
  });

  test('prompts are set', () => {
    const wrapper = shallow(<Prompts {...mockProps} displayPrompts />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().setPrompts = jest.fn();
    wrapper.instance().componentDidMount();
    expect(wrapper.instance().setPrompts).toHaveBeenCalled();
  });
});

describe('method componentDidUpdate()', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Prompts {...mockProps} />);
    wrapper.instance().setPrompts = jest.fn();
  });

  test('calls method setPrompts() while paused is set to false', () => {
    wrapper.setProps({
      displayPrompts: true
    });
    wrapper.update();
    expect(wrapper.instance().setPrompts).toHaveBeenCalled();
  });

  test('calls method setPrompts() while paused is set to true', () => {
    wrapper = shallow(<Prompts {...mockProps} paused />);
    wrapper.instance().setPrompts = jest.fn();
    wrapper.setProps({
      displayPrompts: true,
      paused: false
    });
    wrapper.update();
    expect(wrapper.instance().setPrompts).toHaveBeenCalled();
  });

  test('method setPrompts() is not called', () => {
    wrapper.setProps({
      displayPrompts: false
    });
    wrapper.update();
    expect(wrapper.instance().setPrompts).not.toHaveBeenCalled();
  });
});

test('method onClosePrompt()', () => {
  const wrapper = shallow(<Prompts {...mockProps} />, {
    disableLifecycleMethods: true,
  });
  wrapper.instance().onClosePrompt();
  expect(wrapper.instance().state.displayPrompt).toBeFalsy();
});

describe('method onDisplayPrompt()', () => {
  test('state without any prompts', () => {
    const wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().onDisplayPrompt(0, []);
    expect(wrapper.instance().state).toEqual(expect.objectContaining({
      displayPrompt: true,
      index: 0
    }));
  });

  test('state with prompts', () => {
    const wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().onDisplayPrompt(0, [{ foo: 'bar' }]);
    expect(wrapper.instance().state).toEqual(expect.objectContaining({
      displayPrompt: true,
      index: 0,
      prompts: expect.any(Object)
    }));
  });
});

describe('method onExitPrompt()', () => {
  test('setHintsPaused is called with false', () => {
    // setup
    const wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });

    wrapper.instance().setState({
      index: 1,
      prompts: [{ type: 'foo' }, { type: 'bar' }]
    });

    // function under test
    wrapper.instance().onExitPrompt();

    // expectation
    expect(mockProps.setHintsPaused).toHaveBeenCalledWith(false);
  });

  test('removes prompts and resets the state', () => {
    // setup
    const wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().removePrompts = jest.fn();
    wrapper.instance().setState({
      index: 1,
      prompts: [{ type: 'foo' }, { type: 'bar' }]
    });

    // function under test
    wrapper.instance().onExitPrompt();

    // expectations
    expect(wrapper.instance().removePrompts).toHaveBeenCalledWith(expect.any(String));
    expect(wrapper.instance().state).toEqual(expect.objectContaining({
      index: 0,
      prompts: []
    }));
  });

  test('removes prompts and displays next type prompt', () => {
    // setup
    const wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    const prompts = [{ type: 'foo' }, { type: 'bar' }];
    wrapper.instance().onDisplayPrompt = jest.fn();
    wrapper.instance().removePrompts = jest.fn();
    wrapper.instance().setState({
      index: 0,
      prompts
    });

    // function under test
    wrapper.instance().onExitPrompt();

    // expectations
    expect(wrapper.instance().onDisplayPrompt).toHaveBeenCalledWith(1, prompts);
    expect(wrapper.instance().removePrompts).toHaveBeenCalledWith(expect.any(String));
  });

  test('displays next prompt with the same type', () => {
    // setup
    const wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    const prompts = [{ type: 'foo' }, { type: 'foo' }];
    wrapper.instance().onDisplayPrompt = jest.fn();
    wrapper.instance().removePrompts = jest.fn();
    wrapper.instance().setState({
      index: 0,
      prompts
    });

    // function under test
    wrapper.instance().onExitPrompt();

    // expectations
    expect(wrapper.instance().onDisplayPrompt).toHaveBeenCalledWith(1, prompts);
    expect(wrapper.instance().removePrompts).not.toHaveBeenCalled();
  });
});

describe('method setPrompts()', () => {
  test('no prompts found and summary prompt not shown', () => {
    // setup
    const wrapper = shallow(<Prompts {...mockProps} categories={{}} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().onDisplayPrompt = jest.fn();

    // function under test
    wrapper.instance().setPrompts();

    // expectation
    expect(wrapper.instance().onDisplayPrompt).not.toHaveBeenCalled();
    expect(wrapper.instance().state.summaryPrompt).toBeFalsy();
  });

  test('no prompts found and summary prompt is shown', () => {
    // setup
    const wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().onDisplayPrompt = jest.fn();

    // function under test
    wrapper.instance().setPrompts();

    // expectation
    expect(wrapper.instance().onDisplayPrompt).not.toHaveBeenCalled();
    expect(wrapper.instance().state.summaryPrompt).toBeTruthy();
  });

  test('prompts with achievements', () => {
    // setup
    mockProps.achievements = ['ac1'];
    const wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().onDisplayPrompt = jest.fn();

    // function under test
    wrapper.instance().setPrompts();

    // expectation
    expect(wrapper.instance().onDisplayPrompt).toHaveBeenCalledWith(
      0,
      expect.arrayContaining([expect.objectContaining({
        id: expect.any(String),
        type: ACHIEVEMENT_PROMPT
      })])
    );
  });

  test('prompts with ranks', () => {
    // setup
    mockProps.ranks = ['r1'];
    const wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().onDisplayPrompt = jest.fn();

    // function under test
    wrapper.instance().setPrompts();

    // expectation
    expect(wrapper.instance().onDisplayPrompt).toHaveBeenCalledWith(
      0,
      expect.arrayContaining([expect.objectContaining({
        id: expect.any(String),
        type: RANK_PROMPT
      })])
    );
  });

  test('prompts with a product', () => {
    // setup
    mockProps.products = ['p3'];
    const wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().onDisplayPrompt = jest.fn();

    // function under test
    wrapper.instance().setPrompts();

    // expectation
    expect(wrapper.instance().onDisplayPrompt).toHaveBeenCalledWith(
      0,
      expect.arrayContaining([expect.objectContaining({
        id: expect.any(String),
        family: expect.any(String),
        map: expect.any(String),
        type: PRODUCT_PROMPT,
      })])
    );
  });

  test('bonus streak prompt', () => {
    // setup
    mockProps.bonusStreak = 1;
    const wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().onDisplayPrompt = jest.fn();

    // function under test
    wrapper.instance().setPrompts();

    // expectation
    expect(wrapper.instance().onDisplayPrompt).toHaveBeenCalledWith(
      0,
      expect.arrayContaining([expect.objectContaining({
        type: BONUS_STREAK_PROMPT
      })])
    );
  });

  test('setHintsPaused is called with true', () => {
    // setup
    mockProps.achievements = ['ac1'];
    const wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().onDisplayPrompt = jest.fn();

    // function under test
    wrapper.instance().setPrompts();

    // expectation
    expect(mockProps.setHintsPaused).toHaveBeenCalledWith(true);
  });
});

test('method closeSummaryPrompt()', () => {
  const wrapper = shallow(<Prompts {...mockProps} />, {
    disableLifecycleMethods: true,
  });
  wrapper.instance().closeSummaryPrompt();
  expect(wrapper.instance().state.summaryPrompt).toBeFalsy();
});

describe('method displayPrompts()', () => {
  test('no prompts found', () => {
    const wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().displayPrompts();
    expect(wrapper.instance().state.displayPrompt).toBeFalsy();
  });

  test('displays prompts', () => {
    const wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });
    wrapper.instance().setState({ prompts: ['foo'] });
    wrapper.instance().displayPrompts();
    expect(wrapper.instance().state.displayPrompt).toBeTruthy();
  });
});

describe('method removePrompts()', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Prompts {...mockProps} />, {
      disableLifecycleMethods: true,
    });
  });

  test('no prompts are removed', () => {
    wrapper.instance().removePrompts('');
    expect(mockProps.resetAchievementPrompts).not.toHaveBeenCalled();
    expect(mockProps.resetRankPrompt).not.toHaveBeenCalled();
  });

  test('all achievements prompts are removed', () => {
    wrapper.instance().removePrompts(ACHIEVEMENT_PROMPT);
    expect(mockProps.resetAchievementPrompts).toHaveBeenCalled();
    expect(mockProps.resetProductPrompt).not.toHaveBeenCalled();
    expect(mockProps.resetRankPrompt).not.toHaveBeenCalled();
    expect(mockProps.resetBonusStreakPrompt).not.toHaveBeenCalled();
  });

  test('all product prompts are removed', () => {
    wrapper.instance().removePrompts(PRODUCT_PROMPT);
    expect(mockProps.resetAchievementPrompts).not.toHaveBeenCalled();
    expect(mockProps.resetProductPrompt).toHaveBeenCalled();
    expect(mockProps.resetRankPrompt).not.toHaveBeenCalled();
    expect(mockProps.resetBonusStreakPrompt).not.toHaveBeenCalled();
  });

  test('all rank prompts are removed', () => {
    wrapper.instance().removePrompts(RANK_PROMPT);
    expect(mockProps.resetAchievementPrompts).not.toHaveBeenCalled();
    expect(mockProps.resetProductPrompt).not.toHaveBeenCalled();
    expect(mockProps.resetRankPrompt).toHaveBeenCalled();
    expect(mockProps.resetBonusStreakPrompt).not.toHaveBeenCalled();
  });

  test('bonus streak prompt removed', () => {
    wrapper.instance().removePrompts(BONUS_STREAK_PROMPT);
    expect(mockProps.resetAchievementPrompts).not.toHaveBeenCalled();
    expect(mockProps.resetProductPrompt).not.toHaveBeenCalled();
    expect(mockProps.resetRankPrompt).not.toHaveBeenCalled();
    expect(mockProps.resetBonusStreakPrompt).toHaveBeenCalled();
  });
});

test('should mapStateToProps', () => {
  // setup
  const state = {
    mapData: {
      currentMap: '',
      currentFamily: '',
    },
    promptsData: {
      achievements: [],
      bonusStreak: 0,
      products: [],
      paused: false,
      ranks: [],
      show: false
    },
    scoreData: {
      categories: {}
    }
  };
  const expected = {
    achievements: state.promptsData.achievements,
    bonusStreak: state.promptsData.bonusStreak,
    displayPrompts: state.promptsData.show,
    paused: state.promptsData.paused,
    productFamily: state.mapData.currentFamily,
    productMap: state.mapData.currentMap,
    products: state.promptsData.products,
    ranks: state.promptsData.ranks,
    ...state.scoreData
  };

  // function under test
  const props = mapStateToProps(state);

  // expectation
  expect(props).toEqual(expected);
});

describe('describe mapDispatchToProps', () => {
  test('RESET_ACHIEVEMENT_PROMPTS', () => {
    const dispatch = jest.fn();
    jest.spyOn(promptsActions, 'resetAchievementPrompts')
      .mockReturnValue('resetAchievementPrompts()');
    const props = mapDispatchToProps(dispatch);
    props.resetAchievementPrompts();
    expect(dispatch).toHaveBeenCalledWith({
      type: promptsTypes.RESET_ACHIEVEMENT_PROMPTS
    });
  });

  test('SET_HINTS_PAUSED', () => {
    const dispatch = jest.fn();
    jest.spyOn(hintsActions, 'setHintsPaused').mockReturnValue('setHintsPaused()');
    const props = mapDispatchToProps(dispatch);
    props.setHintsPaused();
    expect(dispatch).toHaveBeenCalledWith({
      type: hintTypes.SET_HINTS_PAUSED
    });
  });
});
