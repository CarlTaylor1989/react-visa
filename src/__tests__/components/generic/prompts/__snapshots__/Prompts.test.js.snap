// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a closed popup 1`] = `
<div
  className="prompts"
>
  <PromptsPopup
    closePopup={[Function]}
    exitPopup={[Function]}
    index={0}
    prompts={Array []}
    setPlayAudio={[MockFunction]}
    showPopup={false}
  />
</div>
`;
